File: jas/algorithms/knuth_plass.py
"""Knuth-Plass every-line line-breaking composer.

Pure-Python port of the Rust / Swift / OCaml KP composers from
``jas_dioxus/src/algorithms/knuth_plass.rs`` etc. Phase 10.

Items are tuples in the form ``("box", width, char_idx)``,
``("glue", width, stretch, shrink, char_idx)``, or
``("penalty", width, value, flagged, char_idx)``. Callers tokenise
their text into items, run :func:`compose`, and re-render lines with
the per-line adjustment ratio applied to glue widths.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


PENALTY_INFINITY = 10000.0


@dataclass(frozen=True)
class KPBox:
    width: float
    char_idx: int


@dataclass(frozen=True)
class KPGlue:
    width: float
    stretch: float
    shrink: float
    char_idx: int


@dataclass(frozen=True)
class KPPenalty:
    width: float
    value: float
    flagged: bool
    char_idx: int


KPItem = KPBox | KPGlue | KPPenalty


def item_width(it: KPItem) -> float:
    if isinstance(it, KPPenalty):
        return 0.0  # contributes only on break
    return it.width


def item_char_idx(it: KPItem) -> int:
    return it.char_idx


@dataclass
class KPOpts:
    line_penalty: float = 10.0
    flagged_demerit: float = 3000.0
    max_ratio: float = 10.0


@dataclass(frozen=True)
class KPBreak:
    item_idx: int
    ratio: float
    flagged: bool


def compose(items: list[KPItem], line_widths: list[float],
            opts: KPOpts | None = None) -> Optional[list[KPBreak]]:
    """Run the Knuth-Plass DP composer.

    ``line_widths`` reuses its last element when the paragraph wants
    more lines than the list provides. Returns ``None`` when no
    feasible composition exists (caller falls back to greedy
    first-fit).

    The returned breaks always end at the final item, which the
    caller must terminate with a forced penalty
    (``value = -PENALTY_INFINITY``).
    """
    if not items or not line_widths:
        return []
    if opts is None:
        opts = KPOpts()
    n = len(items)

    # Prefix sums of width / stretch / shrink for O(1) line eval.
    sum_w = [0.0] * (n + 1)
    sum_y = [0.0] * (n + 1)
    sum_z = [0.0] * (n + 1)
    for i, it in enumerate(items):
        sum_w[i + 1] = sum_w[i] + item_width(it)
        if isinstance(it, KPGlue):
            sum_y[i + 1] = sum_y[i] + it.stretch
            sum_z[i + 1] = sum_z[i] + it.shrink
        else:
            sum_y[i + 1] = sum_y[i]
            sum_z[i + 1] = sum_z[i]

    # DP node: (item_idx, line, total_demerits, ratio, flagged, prev_idx)
    nodes: list[tuple[int, int, float, float, bool, Optional[int]]] = [
        (0, 0, 0.0, 0.0, False, None)
    ]

    def nat_width(from_: int, to_: int) -> tuple[float, float, float]:
        w = sum_w[to_ + 1] - sum_w[from_]
        y = sum_y[to_ + 1] - sum_y[from_]
        z = sum_z[to_ + 1] - sum_z[from_]
        last = items[to_]
        if isinstance(last, KPGlue):
            w -= last.width
            y -= last.stretch
            z -= last.shrink
        elif isinstance(last, KPPenalty):
            w += last.width
        return (w, y, z)

    def line_width_for(line: int) -> float:
        return line_widths[line] if line < len(line_widths) else line_widths[-1]

    for j in range(n):
        item_j = items[j]
        if isinstance(item_j, KPGlue):
            legal = j > 0 and isinstance(items[j - 1], KPBox)
        elif isinstance(item_j, KPPenalty):
            legal = item_j.value < PENALTY_INFINITY
        else:
            legal = False
        if not legal:
            continue

        best: tuple[int, float, float] | None = None
        for ni, n_node in enumerate(nodes):
            n_item_idx, n_line, n_total_d, _, n_flagged, n_prev = n_node
            from_ = 0 if (n_prev is None and ni == 0) else n_item_idx + 1
            if from_ > j:
                continue
            nat, stretch, shrink = nat_width(from_, j)
            line_w = line_width_for(n_line)
            if abs(nat - line_w) < 1e-9:
                ratio = 0.0
            elif nat < line_w:
                ratio = (line_w - nat) / stretch if stretch > 0 else float("inf")
            else:
                ratio = (line_w - nat) / shrink if shrink > 0 else float("-inf")
            if ratio < -1.0 or ratio > opts.max_ratio:
                continue
            badness = 100.0 * abs(ratio) ** 3
            if isinstance(item_j, KPPenalty):
                pen_value, pen_flagged = item_j.value, item_j.flagged
            else:
                pen_value, pen_flagged = 0.0, False
            if pen_value >= 0:
                line_demerit = (opts.line_penalty + badness + pen_value) ** 2
            elif pen_value > -PENALTY_INFINITY:
                line_demerit = (opts.line_penalty + badness) ** 2 - pen_value ** 2
            else:
                line_demerit = (opts.line_penalty + badness) ** 2
            demerits = n_total_d + line_demerit
            if n_flagged and pen_flagged:
                demerits += opts.flagged_demerit
            if best is None or demerits < best[1]:
                best = (ni, demerits, ratio)
        if best is not None:
            prev, d, r = best
            pen_flagged = isinstance(item_j, KPPenalty) and item_j.flagged
            nodes.append((j, nodes[prev][1] + 1, d, r, pen_flagged, prev))

    # Find lowest-demerit node ending at item n-1.
    best_d = float("inf")
    best_idx: Optional[int] = None
    for ni, node in enumerate(nodes):
        if node[0] == n - 1 and node[2] < best_d:
            best_d = node[2]
            best_idx = ni
    if best_idx is None:
        return None
    out: list[KPBreak] = []
    cur = best_idx
    while True:
        node = nodes[cur]
        if node[5] is None and cur == 0:
            break
        out.append(KPBreak(item_idx=node[0], ratio=node[3], flagged=node[4]))
        if node[5] is None:
            break
        cur = node[5]
    out.reverse()
    return out

File: jas_ocaml/lib/algorithms/knuth_plass.ml
(** Knuth-Plass every-line line-breaking composer.

    Pure-OCaml port of [jas_dioxus/src/algorithms/knuth_plass.rs]
    and [JasSwift/Sources/Algorithms/KnuthPlass.swift]. Phase 10. *)

(** One item in the paragraph stream. KP models text as alternating
    boxes (immutable glyph clusters), glue (stretchable / shrinkable
    inter-word space), and penalties (potential break points with
    an associated cost). *)
type item =
  | Box of { width : float; char_idx : int }
  | Glue of { width : float; stretch : float; shrink : float; char_idx : int }
  | Penalty of { width : float; value : float; flagged : bool; char_idx : int }

let item_width = function
  | Box { width; _ } -> width
  | Glue { width; _ } -> width
  | Penalty _ -> 0.0  (* contributes only on break *)

let item_char_idx = function
  | Box { char_idx; _ }
  | Glue { char_idx; _ }
  | Penalty { char_idx; _ } -> char_idx

(** Composer tuning. Defaults match Knuth's original paper. *)
type opts = {
  line_penalty : float;
  flagged_demerit : float;
  max_ratio : float;
}

let default_opts = {
  line_penalty = 10.0;
  flagged_demerit = 3000.0;
  max_ratio = 10.0;
}

(** Penalty value above which a candidate is treated as forbidden. *)
let penalty_infinity = 10000.0

(** One line's break decision. *)
type brk = {
  item_idx : int;
  ratio : float;
  flagged : bool;
}

(** Run the Knuth-Plass DP composer.

    [items] is the linear paragraph stream. [line_widths] reuses
    its last element when the paragraph wants more lines than the
    array provides. Returns [None] when no feasible composition
    exists (caller falls back to greedy first-fit).

    The returned breaks always end at the final item, which the
    caller must terminate with a forced penalty
    ([Penalty { value = -.penalty_infinity; ... }]). *)
let compose ?(opts = default_opts) (items : item array)
    (line_widths : float array) : brk list option =
  let n = Array.length items in
  if n = 0 || Array.length line_widths = 0 then Some []
  else begin
    (* Prefix sums of width / stretch / shrink for O(1) line eval. *)
    let sum_w = Array.make (n + 1) 0.0 in
    let sum_y = Array.make (n + 1) 0.0 in
    let sum_z = Array.make (n + 1) 0.0 in
    for i = 0 to n - 1 do
      sum_w.(i + 1) <- sum_w.(i) +. item_width items.(i);
      (match items.(i) with
       | Glue { stretch; shrink; _ } ->
         sum_y.(i + 1) <- sum_y.(i) +. stretch;
         sum_z.(i + 1) <- sum_z.(i) +. shrink
       | _ ->
         sum_y.(i + 1) <- sum_y.(i);
         sum_z.(i + 1) <- sum_z.(i))
    done;

    (* DP node: a candidate "we broke at this item" record. *)
    let nodes : (int * int * float * float * bool * int option) array ref =
      ref [|
        (* item_idx, line, total_demerits, ratio, flagged, prev *)
        (0, 0, 0.0, 0.0, false, None)
      |] in

    let nat_width from to_ =
      let w = ref (sum_w.(to_ + 1) -. sum_w.(from)) in
      let y = ref (sum_y.(to_ + 1) -. sum_y.(from)) in
      let z = ref (sum_z.(to_ + 1) -. sum_z.(from)) in
      (match items.(to_) with
       | Glue { width; stretch; shrink; _ } ->
         w := !w -. width; y := !y -. stretch; z := !z -. shrink
       | Penalty { width; _ } ->
         w := !w +. width
       | Box _ -> ());
      (!w, !y, !z)
    in

    let line_width_for line =
      let lw = Array.length line_widths in
      if line < lw then line_widths.(line)
      else line_widths.(lw - 1)
    in

    for j = 0 to n - 1 do
      let legal =
        match items.(j) with
        | Glue _ ->
          j > 0 && (match items.(j - 1) with Box _ -> true | _ -> false)
        | Penalty { value; _ } -> value < penalty_infinity
        | Box _ -> false
      in
      if legal then begin
        let best : (int * float * float) option ref = ref None in
        let cur_nodes = !nodes in
        Array.iteri (fun ni (item_idx, line, total_d, _r, flagged, prev) ->
          let from =
            if prev = None && ni = 0 then 0 else item_idx + 1 in
          if from <= j then begin
            let (nat, stretch, shrink) = nat_width from j in
            let line_w = line_width_for line in
            let ratio =
              if Float.abs (nat -. line_w) < 1e-9 then 0.0
              else if nat < line_w then
                if stretch > 0.0 then (line_w -. nat) /. stretch
                else infinity
              else
                if shrink > 0.0 then (line_w -. nat) /. shrink
                else neg_infinity
            in
            if ratio >= -1.0 && ratio <= opts.max_ratio then begin
              let badness = 100.0 *. (Float.abs ratio) ** 3.0 in
              let (pen_value, pen_flagged) = match items.(j) with
                | Penalty { value; flagged; _ } -> (value, flagged)
                | _ -> (0.0, false)
              in
              let line_demerit =
                if pen_value >= 0.0 then
                  (opts.line_penalty +. badness +. pen_value) ** 2.0
                else if pen_value > -. penalty_infinity then
                  (opts.line_penalty +. badness) ** 2.0 -. pen_value ** 2.0
                else
                  (opts.line_penalty +. badness) ** 2.0
              in
              let demerits =
                let base = total_d +. line_demerit in
                if flagged && pen_flagged
                then base +. opts.flagged_demerit
                else base
              in
              (match !best with
               | None -> best := Some (ni, demerits, ratio)
               | Some (_, d_best, _) when demerits < d_best ->
                 best := Some (ni, demerits, ratio)
               | _ -> ())
            end
          end
        ) cur_nodes;
        (match !best with
         | Some (prev, d, r) ->
           let (_, prev_line, _, _, _, _) = cur_nodes.(prev) in
           let pen_flagged =
             match items.(j) with
             | Penalty { flagged; _ } -> flagged
             | _ -> false
           in
           let new_node = (j, prev_line + 1, d, r, pen_flagged, Some prev) in
           nodes := Array.append cur_nodes [| new_node |]
         | None -> ())
      end
    done;

    (* Find lowest-demerit node ending at item n-1. *)
    let best = ref None in
    let best_d = ref infinity in
    Array.iteri (fun ni (item_idx, _, total_d, _, _, _) ->
      if item_idx = n - 1 && total_d < !best_d then begin
        best_d := total_d;
        best := Some ni
      end
    ) !nodes;

    match !best with
    | None -> None
    | Some start ->
      let out = ref [] in
      let cur = ref start in
      let stop = ref false in
      while not !stop do
        let (item_idx, _, _, ratio, flagged, prev) = (!nodes).(!cur) in
        if prev = None && !cur = 0 then stop := true
        else begin
          out := { item_idx; ratio; flagged } :: !out;
          (match prev with
           | Some p -> cur := p
           | None -> stop := true)
        end
      done;
      Some !out  (* already in correct order: head=first break *)
  end
